README.md
SJYouTubeIDParser
=================

A small useful library to extract or parse out the YouTube Video ID from any YouTube URL. You can also use it to fetch information about a YouTube video e.g upload date, title, and thumbnail URL etc...

<img src="Screenshots/screenshot.png" width="600" height="331">

## Requirements
- [AFNetworking](https://github.com/AFNetworking/AFNetworking)
- Runs on iOS 6.0 and later
- Runs on OS X 10.8 and later
- YouTube Data [API Key](https://developers.google.com/youtube/v3/getting-started) (For fetching video info)

## Usage

```
#import "SJYouTubeIDParser.h"
#import "AFNetworking.h"

```
## Get Video Identification
```objc
 SJYouTubeIDParser *youtubeVideoID = [SJYouTubeIDParser defaultParser];
    [youtubeVideoID extractYoutubeID:@"YouTube URL GOES HERE"];
    //This will return a 11 characters Video Identification e.g: w87fOAG8fjk

```

You can also use the new block method

```objc
    SJYouTubeIDParser *youtubeParser = [SJYouTubeIDParser defaultParser];
      [youtubeParser getVideoIDWithURL:@"http://www.youtube.com/watch?v=fmTpgSMJ96c" completionHandler:^(NSString *videoID, NSError *error) {
        
        if (error == nil) {
            // No error
            NSLog(@"Video ID is: %@",videoID);
        } else{
            //Handle error (Video is not a valid YouTube URL)
        }
        
    }];
```
## Get information about a YouTube Video

**Note**:You need a YouTube Data [API Key](https://developers.google.com/youtube/v3/getting-started) to use this method.

```objc
//Set API Key
youtubeParser.APIKey = @"YOUR KEY";

[youtubeParser getVideoInfoFromURL:@"http://www.youtube.com/watch?v=fmTpgSMJ96c" completionHandler:^(NSString *videoID, NSString *videoTitle, NSURL *thumbnailURL, NSString *videoDescription, NSString *viewCount, long long likeCount, long long dislikeCount, NSDate *uploaded, NSString *uploader, NSError *error) {
        
        if (error == nil) {
            // No error
            //you can now get YouTube video's
            //video ID, thumbnail URL, video description, view count,like count, dislike count, upload date, and the uploader

        
        } else{
            //Handle error
        }
        
    }];
 ```

## Supported video URLs are:

- http://www.youtube.com/watch?v=sapuE9Cui0g&feature=youtu.be
- http://www.youtube.com/watch?v=sapuE9Cui0g
- youtu.be/KFPtWedl7wg_U923
- http://www.youtube.com/watch?feature=player_detailpage&v=sapuE9Cui0g_U#t=31s

## License

SJYouTubeIDParser is available under the MIT license. See the [LICENSE](https://github.com/SoneeJohn/SJYouTubeIDParser/blob/master/LICENSE.md) file for more information.

CHANGELOG.md
#### Version 1.2.0 

* Updated to YouTube API V3
* Updated Screenshot (OS X Yosemite)
* Add **SJYouTubeIDParserThumbnailQuailty** property
* Added **APIKey** property. 
* **Note**: You need to specify this if you wish to fetch info about YouTube videos
* Updated Demo project to reflect the above changes

#### Version 1.1.2

* Small bug fix.
* Can how report if video is deleted or private.
* Updated Demo Project to reflect on the above changes
* Added MIT License.


#### Version 1.1.1

* Improved the library's usages of ``` NSError```
* Improved error reporting
* Updated Demo Project to reflect on the above changes

#### Version 1.1.0

* Added block methods
* Updated Demo Project
* When calling "getVideoInfoFromURL" you can now get YouTube video's
video ID, thumbnail URL, video description, view count,rating, upload date, and the uploader
* Added + defaultParser instance
* Fixed all known issues
* Now requires AFNetworking


#### Version 1.1.0 bc1

* Added blocks
* Updated Demo Project
* When calling "getVideoInfoFromURL" you can now get YouTube video's
video ID, thumbnail URL, video description, view count, and the rating
* Added + defaultParser instance

**Known issues:**
* Parser will crash if attemping to request video info with no internet
connection.

#### Version 1.0.0

* Initial version


SJYouTubeIDParser Demo/SJYouTubeIDParser Demo/AppDelegate.h
//
//  AppDelegate.h
//  SJYouTubeIDParser Demo
//
//  Created by Soneé John on 7/17/14.
//  Copyright (c) 2014 AlphaSoft. All rights reserved.
//

#import <Cocoa/Cocoa.h>

@interface AppDelegate : NSObject <NSApplicationDelegate>{
    
    __weak NSTextField *_videoIDLabel;
    __weak NSTextField *_textField;

    __weak NSTextField *_videoTitleTextField;

    __weak NSProgressIndicator *_circularProgressIndicator;

    __weak NSTextField *_descriptionTextField;


    __weak NSTextField *_timeAgoTextField;
    __weak NSImageView *_thumbnailImageView;

    __weak NSTextField *_uploaderTextField;
    __weak NSLevelIndicator *_ratingIndicator;

    __weak NSTextField *_viewCountTextField;
}

@property (assign) IBOutlet NSWindow *window;
@property (weak) IBOutlet NSTextField *likesLabel;
@property (weak) IBOutlet NSTextField *dislikesLabel;

@property (weak) IBOutlet NSTextField *textField;
- (IBAction)extractAction:(id)sender;
@property (weak) IBOutlet NSTextField *videoIDLabel;
@property (weak) IBOutlet NSTextField *videoTitleTextField;
@property (weak) IBOutlet NSTextField *descriptionTextField;
@property (weak) IBOutlet NSImageView *thumbnailImageView;
@property (weak) IBOutlet NSTextField *viewCountTextField;
@property (weak) IBOutlet NSLevelIndicator *ratingIndicator;
@property (weak) IBOutlet NSProgressIndicator *circularProgressIndicator;
@property (weak) IBOutlet NSTextField *timeAgoTextField;
@property (weak) IBOutlet NSTextField *uploaderTextField;
@end


SJYouTubeIDParser Demo/SJYouTubeIDParser Demo/AppDelegate.m
//
//  AppDelegate.m
//  SJYouTubeIDParser Demo
//
//  Created by Soneé John on 7/17/14.
//  Copyright (c) 2014 AlphaSoft. All rights reserved.
//

#import "AppDelegate.h"
#import "SJYouTubeIDParser.h"
@implementation AppDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification
{
    // Insert code here to initialize your application
}



- (IBAction)extractAction:(id)sender {
    
    
    //Create SJYouTubeIDParser object
   SJYouTubeIDParser *youtubeParser = [SJYouTubeIDParser defaultParser];
    //Start spinner
    //Undide it first
    [self.circularProgressIndicator setHidden:NO];
  [self.circularProgressIndicator startAnimation:nil];
    
    youtubeParser.APIKey = @"AIzaSyCTu9h7PfHd7LI0Wy7Ep4VY4tM40fWOxb4";
    [youtubeParser getVideoInfoFromURL:[self.textField stringValue] completionHandler:^(NSString *videoID, NSString *videoTitle, NSURL *thumbnailURL, NSString *videoDescription, NSString *viewCount, long long likeCount, long long dislikeCount, NSDate *uploaded, NSString *uploader, NSError *error) {
        
        [self.circularProgressIndicator stopAnimation:nil];

        if (error == nil) {
            // No error
            //you can now get YouTube video's
            //video ID, thumbnail URL, video description, view count,rating, upload date, and the uploader
            
            //Title
            [self.videoTitleTextField setStringValue:videoTitle];

            
            [self.descriptionTextField setStringValue:videoDescription];
            
            NSImage *thumbnailImage = [[NSImage alloc]initByReferencingURL:thumbnailURL];
            //Set thumbnail
            [self.thumbnailImageView setImage:thumbnailImage];
            
            //Uploader
            [self.uploaderTextField setStringValue:[NSString stringWithFormat:@"by %@",uploader]];
            
            
            //Views
            NSNumberFormatter *formatter = [[NSNumberFormatter alloc] init];
            [formatter setNumberStyle:NSNumberFormatterDecimalStyle];
            NSString *groupingSeparator = [[NSLocale currentLocale] objectForKey:NSLocaleGroupingSeparator];
            [formatter setGroupingSeparator:groupingSeparator];
            [formatter setGroupingSize:3];
            [formatter setAlwaysShowsDecimalSeparator:NO];
            [formatter setUsesGroupingSeparator:YES];
            [self.viewCountTextField setStringValue:[NSString stringWithFormat:@"%@ views",[formatter stringFromNumber:[NSNumber numberWithInteger:[viewCount integerValue]]]]];
            
            //Likes
            
            [self.likesLabel setStringValue:[NSString stringWithFormat:@"%@ likes",[formatter stringFromNumber:[NSNumber numberWithLongLong:likeCount]]]];

             
             //Dislikes
             [self.dislikesLabel setStringValue:[NSString stringWithFormat:@"%@ dislikes",[formatter stringFromNumber:[NSNumber numberWithLongLong:dislikeCount]]]];

            
            //Uploaded
            
            NSDate * today = [NSDate date];
            
            NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
            [dateFormatter setDateFormat:@"yyyy-MM-dd"];
            NSString *strDate = [dateFormatter stringFromDate:uploaded];
            
            NSString * str = strDate;
            NSDate * past = [NSDate dateWithNaturalLanguageString:str
                                                           locale:[[NSUserDefaults
                                                                    standardUserDefaults] dictionaryRepresentation]];
            
            
            NSCalendar *gregorian = [[NSCalendar alloc]
                                     initWithCalendarIdentifier:NSGregorianCalendar];
            unsigned int unitFlags = NSYearCalendarUnit | NSMonthCalendarUnit |
            NSDayCalendarUnit |
            NSHourCalendarUnit | NSMinuteCalendarUnit |
            NSSecondCalendarUnit;
            NSDateComponents *components = [gregorian components:unitFlags
                                                        fromDate:past
                                                          toDate:today
                                                         options:0];
            
            
            if ([components year]) {
                [self.timeAgoTextField setStringValue:[NSString stringWithFormat:@"%ld years ago",(long)[components year]]];
                
            } else if ([components month]){
                [self.timeAgoTextField setStringValue:[NSString stringWithFormat:@"%ld months ago",(long)[components month]]];
                
            } else if ([components day]){
                [self.timeAgoTextField setStringValue:[NSString stringWithFormat:@"%ld days ago",(long)[components day]]];
                
            } else if ([components hour]){
                [self.timeAgoTextField setStringValue:[NSString stringWithFormat:@"%ld hours ago",(long)[components hour]]];
                
                
            } else if ([components minute]){
                [self.timeAgoTextField setStringValue:[NSString stringWithFormat:@"%ld minutes ago",(long)[components minute]]];
                
                
            } else if ([components second]){
                [self.timeAgoTextField setStringValue:[NSString stringWithFormat:@"%ld seconds ago",(long)[components second]]];
                
                
            }
        
        } else{
            //Handle error
            
            if ([error code] == 10) {
                NSLog(@"Invalid URL");
                NSRunAlertPanel(NSLocalizedString(@"Invalid URL", @"Invalid URL alert title"), NSLocalizedString(@"Unable to find any YouTube video with the URL you provided. Your URL should look something like this: http://www.youtube.com/watch?v=VpZmIiIXuZ0", @"Message for invalid URL ALert"), NSLocalizedString(@"OK", @"OK button for invalid URL alert"), nil, nil);
                
            } else if ([error code] == 20){
                NSLog(@"Video is deleted");
                [NSApp presentError:error];
                
                
            } else if ([error code] == 30){
                NSLog(@"Video is private");
                [NSApp presentError:error];
                
            } else{
                
                
                [NSApp presentError:error];
                
            }

        }
        
    }];
    
    
    
}
@end


SJYouTubeIDParser Demo/SJYouTubeIDParser Demo/SJYouTubeIDParser.h
//
//  SJYouTubeIDParser.h
//  SJYouTubeIDParser
//
//  Created by Soneé John on 7/17/14.
//  Copyright (c) 2014 AlphaSoft. All rights reserved.
//

//**** License ****//
/*
The MIT License (MIT)

Copyright (c) 2015 Soneé John

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#import <Foundation/Foundation.h>
@interface SJYouTubeIDParser : NSObject

typedef NS_ENUM(NSUInteger, SJYouTubeIDParserThumbnailQuailty) {
    
    SJYouTubeIDParserThumbnailQuailtyDefault  = 1,
    
    SJYouTubeIDParserThumbnailQuailtyMedium = 2,
    
    SJYouTubeIDParserThumbnailQuailtyHigh = 3,
};
/**
 * The thumbnail quailty for the YouTube video.
 
  @b Note @b: Default value is, @b SJYouTubeIDParserThumbnailQuailtyHigh @b
 */
@property (nonatomic, assign) SJYouTubeIDParserThumbnailQuailty thumbnailQuailty;


//** The following is can example of a deleted YouTube video any attempts to fetch info about the video will result in a (404) error *//
//http://www.youtube.com/watch?v=vsxCwsD9HoI&feature=related.

/**
 Method for retrieving the youtube ID from a youtube URL
 
 @param youtubeURL the the complete youtube video url, either youtu.be or youtube.com
 @return the YouTube ID
 
 Supported video URLs are:
 http://www.youtube.com/watch?v=sapuE9Cui0g&feature=youtu.be
 http://www.youtube.com/watch?v=sapuE9Cui0g
 youtu.be/KFPtWedl7wg_U923
 http://www.youtube.com/watch?feature=player_detailpage&v=sapuE9Cui0g_U#t=31s
 */

- (NSString *)extractYoutubeID:(NSString *)youtubeURL;
+ (instancetype) defaultParser;
/**
 * The API Key that is required to fetch information about YouTube videos
 */
@property (nonatomic, assign) NSString *APIKey;


/**
 *  Gets the video id from a specified video url, and calls a handler upon completion.
 */
- (void) getVideoIDWithURL:(NSString *)youtubeURL completionHandler:(void (^)(NSString *videoID, NSError *error))completionHandler;

/**
 *  Fetches video info from the specified video url, and calls a handler upon completion.
 
 */
- (void) getVideoInfoFromURL:(NSString *)youtubeURL completionHandler:(void (^)(NSString *videoID, NSString *videoTitle, NSURL *thumbnailURL, NSString *videoDescription, NSString *viewCount, long long likeCount, long long dislikeCount, NSDate *uploaded,NSString *uploader, NSError *error))completionHandler;

@end


SJYouTubeIDParser Demo/SJYouTubeIDParser Demo/SJYouTubeIDParser.m
//
//  SJYouTubeIDParser.m
//  SJYouTubeIDParser
//
//  Created by Soneé John on 7/17/14.
//  Copyright (c) 2014 AlphaSoft. All rights reserved.
//

//**** License ****//
/*
 The MIT License (MIT)
 
 Copyright (c) 2015 Soneé John
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

#import "SJYouTubeIDParser.h"
#import "AFNetworking.h"
@interface SJYouTubeIDParser ()


@end

@implementation SJYouTubeIDParser

+ (instancetype) defaultParser
{
	static SJYouTubeIDParser *defaultParser;
	static dispatch_once_t once;
	dispatch_once(&once, ^{
		defaultParser = [self new];

	});
	return defaultParser;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        
        self.thumbnailQuailty = 3;
    }
    return self;
}

- (NSString *)extractYoutubeID:(NSString *)youtubeURL
{
    NSError *error = NULL;
    
    //Create a NSRegularExpression object
    NSRegularExpression *regexExp = [NSRegularExpression regularExpressionWithPattern:@"(?<=v(=|/))([-a-zA-Z0-9_]+)|(?<=youtu.be/)([-a-zA-Z0-9_]+)" options:NSRegularExpressionCaseInsensitive error:&error];
    
    NSRange rangeOfFirstMatch = [regexExp rangeOfFirstMatchInString:youtubeURL options:0 range:NSMakeRange(0, [youtubeURL length])];
    if(!NSEqualRanges(rangeOfFirstMatch, NSMakeRange(NSNotFound, 0)))
    {
        //Matches
        NSString *videoIdentification = [youtubeURL substringWithRange:rangeOfFirstMatch];
        
        //Return Video ID
        return videoIdentification;
        
    } else{
        
        //Error
        NSError *error = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:10 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"Invalid YouTube URL" }];

        NSLog(@"SJYouTubeIDParser returned error: %@",error);
    }
    return nil;
}
- (void) getVideoIDWithURL:(NSString *)youtubeURL completionHandler:(void (^)(NSString *videoID, NSError *error))completionHandler{
    NSError *error = NULL;
    //Create a NSRegularExpression object
    NSRegularExpression *regexExp = [NSRegularExpression regularExpressionWithPattern:@"(?<=v(=|/))([-a-zA-Z0-9_]+)|(?<=youtu.be/)([-a-zA-Z0-9_]+)" options:NSRegularExpressionCaseInsensitive error:&error];
    
    NSRange rangeOfFirstMatch = [regexExp rangeOfFirstMatchInString:youtubeURL options:0 range:NSMakeRange(0, [youtubeURL length])];
    if(!NSEqualRanges(rangeOfFirstMatch, NSMakeRange(NSNotFound, 0)))
    {
        //Matches
        //Return Video ID
        NSString *videoIdentification = [youtubeURL substringWithRange:rangeOfFirstMatch];
        completionHandler(videoIdentification, error);
        


    } else{
        
        //Error
        NSError *error = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:10 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"Invalid YouTube URL" }];

        completionHandler(nil, error);

    }

}
- (void) getVideoInfoFromURL:(NSString *)youtubeURL completionHandler:(void (^)(NSString *videoID, NSString *videoTitle, NSURL *thumbnailURL, NSString *videoDescription, NSString *viewCount, long long likeCount, long long dislikeCount, NSDate *uploaded,NSString *uploader, NSError *error))completionHandler{
    
    if([self.APIKey length] >0){
        
   
   
    NSError *error = NULL;
    //Create a NSRegularExpression object
    NSRegularExpression *regexExp = [NSRegularExpression regularExpressionWithPattern:@"(?<=v(=|/))([-a-zA-Z0-9_]+)|(?<=youtu.be/)([-a-zA-Z0-9_]+)" options:NSRegularExpressionCaseInsensitive error:&error];
    
    NSRange rangeOfFirstMatch = [regexExp rangeOfFirstMatchInString:youtubeURL options:0 range:NSMakeRange(0, [youtubeURL length])];
    
    //Found the Video ID This mean that this is a valid YouTube URL
    if(!NSEqualRanges(rangeOfFirstMatch, NSMakeRange(NSNotFound, 0)))
    {
        
                NSString *videoIdentification = [youtubeURL substringWithRange:rangeOfFirstMatch];
                
                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0)
                               , ^(void) {
                                   // Base URL to get video info on any YouTube Video:
                                   // http://gdata.youtube.com/feeds/api/videos/<Video ID HERE>?v=2&alt=jsonc
                                   
                                   //..1
                                   
                               //https://www.googleapis.com/youtube/v3/videos?part=snippet%@+statistics&id=%@&key=%@

                                   
                                   NSString *constString = @"%2C";
                                   NSString *jsonString = [NSString stringWithFormat:@"https://www.googleapis.com/youtube/v3/videos?part=snippet%@+statistics&id=%@&key=%@",constString,videoIdentification, self.APIKey];
                                   
                                   // Fetch Json info
                                    AFHTTPRequestOperationManager *manager = [AFHTTPRequestOperationManager manager];
                                   [manager GET:jsonString parameters:nil success:^(AFHTTPRequestOperation *operation, id responseObject) {
                                       //parse out the json data
                                       
                                       NSDictionary *dict = (NSDictionary *)responseObject;

                                       NSMutableDictionary *parseResults = [[NSMutableDictionary alloc] initWithDictionary:dict];
                                       
                                       
                                       @try {
                                          // Code that can potentially throw an exception
                                       
                                       
                                       //Get video Title
                                        NSString *youtubeVideoTitle = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"title"];
                                       
                                       
                                       //Get URL for thumbnail
                                      
                                       NSURL *thumbnailURL;
                                       if (self.thumbnailQuailty == 1) {
                                           //default
                                           thumbnailURL = [NSURL URLWithString:[[[[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"thumbnails"] objectForKey:@"default"]valueForKey:@"url"]];//default
                                           
                                       } else if (self.thumbnailQuailty == 2){
                                           //Medium
                                           
                                           thumbnailURL = [NSURL URLWithString:[[[[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"thumbnails"] objectForKey:@"medium"]valueForKey:@"url"]];//Medium
                                           
                                           
                                       } else if (self.thumbnailQuailty == 3){
                                           //High
                                           
                                           thumbnailURL = [NSURL URLWithString:[[[[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"thumbnails"] objectForKey:@"high"]valueForKey:@"url"]];//high
                                           
                                       } else{
                                           
                                           //Fall back to default
                                           
                                           thumbnailURL = [NSURL URLWithString:[[[[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"thumbnails"] objectForKey:@"default"]valueForKey:@"url"]];//default
                                           
                                       }
                                       
                                       
                                       
                                       //Get video Description
                                       NSString *videoDescription = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"description"];
                                       
                                       //Get View Count
                                       NSString *viewCount = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"statistics"] objectForKey:@"viewCount"];
                                       
                                       //Get likeCount
                                       NSString *likeCount = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"statistics"] objectForKey:@"likeCount"];
                                       
                                       NSInteger likeCountInt = [likeCount intValue];
                                       
                                       
                                       //Get dislikeCount
                                       NSString *dislikeCount = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"statistics"] objectForKey:@"dislikeCount"];
                                       
                                       NSInteger dislikeCountInt = [dislikeCount intValue];
                                       
                                       
                                       //Get uploaded date
                                       NSDate *upld = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"publishedAt"];
                                       
                                       NSString *myDate = [[NSString stringWithFormat:@"%@", upld] substringToIndex:10];
                                       
                                       NSDateFormatter* df_utc = [[NSDateFormatter alloc] init];
                                       [df_utc setDateFormat:@"yyyy-MM-dd"];
                                       upld = [df_utc dateFromString:myDate];
                                       
                                       NSDate *uploadedDate = [df_utc dateFromString:myDate];
                                       
                                       
                                       // Get uploader
                                       NSString *uploader = [[[[parseResults objectForKey:@"items"] objectAtIndex:0] objectForKey:@"snippet"] objectForKey:@"channelTitle"];
                                       
                                       
                                       
                                       completionHandler(videoIdentification, youtubeVideoTitle, thumbnailURL, videoDescription,viewCount, likeCountInt,dislikeCountInt,uploadedDate, uploader, error);
                                       }@catch (NSException *exception) {
                                           //Handle an exception thrown in the @try block
                                           
                                           if ([parseResults objectForKey:@"items"] == nil || [[parseResults objectForKey:@"items"]count] == 0) {
                                               
                                               NSError *unavailableError = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:50 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"This video is unavailable."}];
                                               
                                               completionHandler(nil, nil, nil, nil,nil, 0, 0,nil, nil, unavailableError);

                                           } else{
                                           
                                           NSError *unknownError = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:60 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"An known error occurred"}];
                                           
                                           completionHandler(nil, nil, nil, nil,nil, 0, 0,nil, nil, unknownError);
                                               
                                           }

                                       }
                                       

                                       
                                      
                                   } failure:^(AFHTTPRequestOperation *operation, NSError *error) {
                                       //Handle error
                                       NSInteger statusCode = operation.response.statusCode;

                                       
                                       if (statusCode == 404) {
                                           //Video was deleted
                                           NSError *deletedError = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:20 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"The video requested was deleted."}];

                                           completionHandler(nil, nil, nil, nil,nil, 0, 0,nil, nil, deletedError);

                                           
                                       } else if (statusCode == 403){
                                           
                                          //The video was private
                                           NSError *privateError = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:30 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"The video requested is private."}];

                                           completionHandler(nil, nil, nil, nil,nil, 0,0,nil, nil, privateError);
 
                                           
                                       } else{
                                          //some other error
                                           completionHandler(nil, nil, nil, nil,nil, 0,0,nil, nil, error);

                                       }

                                       
                                   }];
                                   
                               });
        
    } else{
        // Invalid URL
        
        NSError *error = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:10 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"Invalid YouTube URL" }];
        NSString *youtubeVideoTitle = nil;
        NSURL *thumbnailURL = nil;
        NSString *videoDescription = nil;
        NSString *viewCount = nil;
        NSInteger likeCountInt;
        NSInteger dislikeCountInt;

        NSString *videoIdentification = nil;
        NSString *uploader = nil;
        NSDate *uploadedDate = nil;
        completionHandler(videoIdentification, youtubeVideoTitle, thumbnailURL, videoDescription,viewCount, likeCountInt,dislikeCountInt,uploadedDate, uploader, error);
    }
    } else{
        
        NSLog(@"Please enter your YouTube API Key");
        
        NSError *error = [NSError errorWithDomain:@"com.AlphaSoft.SJYouTubeIDParser" code:40 userInfo:@{ NSLocalizedFailureReasonErrorKey : @"YouTube API Key Not Specified" }];
        NSString *youtubeVideoTitle = nil;
        NSURL *thumbnailURL = nil;
        NSString *videoDescription = nil;
        NSString *viewCount = nil;
        NSInteger likeCountInt;
        NSInteger dislikeCountInt;
        
        NSString *videoIdentification = nil;
        NSString *uploader = nil;
        NSDate *uploadedDate = nil;
        completionHandler(videoIdentification, youtubeVideoTitle, thumbnailURL, videoDescription,viewCount, likeCountInt,dislikeCountInt,uploadedDate, uploader, error);

            }

}


@end
